File: app/ts/src/auth/modules/api_key_query.ts
import type { AuthContext, AuthResult } from "../../core/types.ts";
import type { AuthModule, AuthDeps } from "../types.ts";
import { invalidateOnFromConfig } from "../types.ts";

export class ApiKeyQueryAuth implements AuthModule {
  static readonly typeName = "api_key_query";

  private readonly token: string;
  private readonly param: string;
  private readonly invalidateOn: Set<number>;

  private constructor(token: string, param: string, invalidateOn: Set<number>) {
    this.token = token;
    this.param = param;
    this.invalidateOn = invalidateOn;
  }

  static fromConfig(
    config: Record<string, unknown>,
    deps: AuthDeps,
  ): ApiKeyQueryAuth {
    if (typeof config["env"] !== "string") {
      throw new Error("ApiKeyQueryAuth: 'env' config field is required.");
    }
    const envName = config["env"];
    const envVal = deps.env[envName];
    if (typeof envVal !== "string" || !envVal) {
      throw new Error(
        `ApiKeyQueryAuth: env var '${envName}' is not set or empty.`,
      );
    }

    if (typeof config["param"] !== "string") {
      throw new Error("ApiKeyQueryAuth: 'param' config field is required.");
    }

    const invalidateOn = invalidateOnFromConfig(config, [401]);

    return new ApiKeyQueryAuth(envVal, config["param"], invalidateOn);
  }

  async apply(_ctx: AuthContext): Promise<AuthResult> {
    return { queryParams: { [this.param]: this.token } };
  }

  async invalidate(): Promise<void> {}

  isRejection(statusCode: number, _body: unknown): boolean {
    return this.invalidateOn.has(statusCode);
  }
}

File: app/python/auth/modules/api_key_query.py
from __future__ import annotations

import os
from typing import Any

import httpx
from redis.asyncio import Redis

from app.python.auth.base import AuthContext, AuthResult, _invalidate_on_from_config


class ApiKeyQueryAuth:
    """API key injected as a query parameter."""

    name = "api_key_query"

    def __init__(
        self,
        token: str,
        param: str,
        invalidate_on: set[int],
    ) -> None:
        self._token = token
        self._param = param
        self._invalidate_on = invalidate_on

    @classmethod
    def from_config(
        cls, config: dict[str, Any], *, redis: Redis, http: httpx.AsyncClient
    ) -> ApiKeyQueryAuth:
        env_name = config["env"]
        token = os.environ.get(env_name)
        if not token:
            raise ValueError(f"ApiKeyQueryAuth: env var {env_name!r} is not set or empty.")
        return cls(
            token=token,
            param=config["param"],
            invalidate_on=_invalidate_on_from_config(config),
        )

    async def apply(self, ctx: AuthContext) -> AuthResult:
        return AuthResult(query_params={self._param: self._token})

    async def invalidate(self) -> None:
        pass

    def is_rejection(self, status_code: int, body: dict | None) -> bool:
        return status_code in self._invalidate_on
